### PR TITLE
use normalized args to silence warnings in DEBUG mode

### PR DESCRIPTION
**Pull request by jjn1056**
The issue with this change is that in the case where Args is a Type Constraint, it results in silly warnings and incorrect output when CATALYST_DEBUG is on.  This is because we use this value to display '*' for the number of arguments and when this is a string you get a 'can't use string in a repeat error (see test case).

We changed this in commit https://github.com/perl-catalyst/catalyst-runtime/commit/bd6018c40240ca651d1ff62552b948fc9e691cab

which was possible done because of the issue when Args is undefined thats the cue for 'infinite' and that lead to an out of memory error when CATALYST_DEBUG was true.  HOWEVER the code used to display the argument count doesn't blindly use the Args count, it checks to see if this is undefined first, so this is no longer a problem (see https://github.com/perl-catalyst/catalyst-runtime/blob/master/lib/Catalyst/DispatchType/Chained.pm#L100)

Basically ->list_extra_info->{Args} needs to be a number or undefined based on its usage.  If we don't want to use ->normalized_arg_number because that can be ~0 we can do something here that tests the attribute to see if its a number or not and does the right thing or not.  Current setup spews warnings all over my logs and messes up the DEBUG display

### CODE REVIEW

**Comment by jjn1056** [comment]
@vanstyn hey, would appreciate a review of the above since you have comaint on the repo.  This is causing me tons of trouble actually.  Please let me know if I can explain better.

**Comment by vanstyn** [comment]
@jjn1056 - seems reasonable, as long as you're confident you understand the reason behind @shadowcat-mst 's bd6018c change and that it is in fact no longer an issue. I've never encountered issues in this area of the code (nor have I messed with it much) so I'm probably not the best person to ask. I trust your judgement, especially since it looks like you've added a proper new test for this, as well as argument/discussions, so if an issue arises in the future it will be easier to see whats going on and what happened.

I went ahead restarted a couple of the failed travis builds for this PR which looked like hiccups so hopefully we can see a thumbs up from travis shortly...

**Comment by perl-catalyst-sync** [comment]
Last time I tried to change this mst forcible overwrote my change.  He
didn't really explain it and to be honest at this point I'm not interested
in hearing the explanation (the time for that was 6 months ago) but you
probably need him to review since its possible he has some obscure but
valid reason for it being this way.  My guess to his change was related to
the out of memory issue but maybe he's got client code that does other
stuff with ->{Args} that I never heard of so can't be sure.  I pinged him
in a message about the commit but he probably doesn't hit github very much
and I no longer wish to use a community service managed by shadowcat, at
least until there's clear community guidelines that people believe in.  So
sorry I'm aiming this at you.  If you'd rather not deal with it I
understand, just trying to get the attention of someone with comaint to
figure this out.  I can always release an ActionRole that fixes this for my
code and leave it at that.  Thanks

Not sure, travis is cranky a bit.  I'll keep an eye on it.  -jnap

**Comment by shadowcat-mst** [comment]
I just overwrote a bunch of bad merges since I was in the middle of rebasing the branches we'd already decided we needed before your IRC connection dropped - I figured you'd realise that once you saw the same branches go in rebased and then once I'd got stuff merged cleanly we'd pick the conversation back up on channel. Then you never rejoined so I never got the chance to explain.

It would seem to me that having the debug code have to handle ~0 is the wrong answer though - aren't we better off having either (a) the things relying on ~0 for search special case undef or (b) have a normalized_arg_number *and* an arg_number? ... Or (c) we could parse Args into [ 'Str', 'Int' ] and have the debug code be even more intelligent, so instead of /*/* you get /*[Str]/*[Int] or something, which strikes me as a rather neat idea. Thoughts?

**Comment by shadowcat-mst** [comment]
Also @perl-catalyst-sync didn't you mean to comment as @jjn1056 ? Not sure how github managed to do that to your comment ...

**Comment by perl-catalyst-sync** [comment]
On Mon, Feb 22, 2016 at 2:26 PM, shadowcat-mst <notifications@github.com>
wrote:
>
> I just overwrote a bunch of bad merges since I was in the middle of
rebasing the branches we'd already decided we needed before your IRC
connection dropped - I figured you'd realise that once you saw the same
branches go in rebased and then once I'd got stuff merged cleanly we'd pick
the conversation back up on channel. Then you never rejoined so I never got
the chance to explain.

ok fair enough, I get that now.  Just right now for my sanity and with all
the other things troubling my personal life I want to be calm and
reflective.  I can't do that interactively on IRC anymore, or at least I
can't right now.  So I can work via 'detached' media like email etc. since
that gives me the chance to put down stuff when I can't deal with it and go
off to do stuff like walk my dogs.  Its likely I'll attend YAPC later this
year and if anyone in the community wants to get together and try to figure
out what the Catalyst goals and roadmap could be I'm happy to hear it,
although I personally feel I'd want more clear guidelines around
conversations and decision making before I could take any active role.

>
> It would seem to me that having the debug code have to handle ~0 is the
wrong answer though - aren't we better off having either (a) the things
relying on ~0 for search special case undef or (b) have a
normalized_arg_number and an arg_number? ... Or (c) we could parse Args
into [ 'Str', 'Int' ] and have the debug code be even more intelligent, so
instead of // you get /[Str]/[Int] or something, which strikes me as a
rather neat idea. Thoughts?

I'm not understanding, "(a) the things relying on ~0 for search special
case undef ".   Do you mean that we should use undef instead of ~0  and the
code that does the request matching handling the 'its unbounded'?  In other
words normalized_arg_number normalized to undef not ~0?

I agree having 'sideband' information that ~0 needs to be special handled
is ugly and error prone, its not a good API when the only way to understand
it is "I heard from a guy at the pub".  Right now we have to have some wack
code that looks to see if attributes->{Args}[0] is undef, which is the
sideband info to say, 'ignore list_extra_info->{Args} for the purposes of
display, its too big, use the ... to indicate infinite...

However I was really worried about changing the nature of what
->llist_extra_info->{Args} returned.  So I decided to leave it returning a
number.  I can't find it now, but I did find at least one case on CPAN that
had code that used that to do something, I think generate docs or auto
generate API (for some reason I though there was a Swagger generator that
did this but now I can't find that).  So if option A is to return to the
idea of undef means infinite perhaps that is the approach least likely to
bust stuff.

It would be great if the Path Spec column in the DEBUG screen could show
the constraints,  although I'm not sure the best way to render Tuples with
that (and Tuple is allowed).  The main reason for allowing a Tuple was I
wanted to introduce a way to conceptually link several arguments together
as all inter-related from the viewpoint of a constraint.  Perhaps that's
not such a useful idea (felt useful to me at the time).

If option A means 'ditch normalizing to ~0, normalize infinity to undef and
refactor the debug display code and the matching code based on that, I can
see that.  Please let me know if I am not clear on it.

-john

**Comment by perl-catalyst-sync** [comment]
Its wacky because when I commit to github we communicate that to shadowgit
via the special user 'perl-catalyst-sync'.  We never really bothered to
figure out two way communication between shadowgit and github. I believe
its correct in git log ...

**Comment by shadowcat-mst** [comment]
It's not wacky, when you commit to shadowgit, shadowgit pushes up via the perl-catalyst-sync user. I know how that works, I set it up for you if you recall.

What I'm confused by is that you're commenting on this issue through the web interface as perl-catalyst-sync rather than your own user. Surely once you'd set perl-catalyst-sync up, you can log out and never log back into it again, and then you'd be able to comment as you?

**Comment by shadowcat-mst** [comment]
Oh: Two way sync basically isn't worth it because github's 'accept PR' button is unusable - if you want to rebase and merge cleanly, you have to do it from a checkout anyway, and once you've done that it's just as easy to push the result to both the PR and master - and then github auto-closes the PR for you anyway. I recommend ingy's git-hub tool for making this less confusing.

**Comment by jjn1056** [comment]
maybe a bug, lets see who this says I am.  Right you can't accept the PR via the GH interface, you need to pull it down and do it all via the command line, but in the end it gets closed properly.  I'll look at ingy's  stuff, but I do this so often its always a few commands in the .history file anyway.

**Comment by jjn1056** [comment]
ok that one said I was jnap.  weird

**Comment by shadowcat-mst** [comment]
Option A does mean that, though the ~0 thing isn't mine so I still really don't understand where it came from. I managed to get Chained/Path basically working without it, and I don't think it's really the right name *or* the right behaviour.

I think ideally, we'd

1) have a comparable_arg_number method that actually makes sense to have ~0 in
2) have normalized_arg_number return just a normalized value (or undef), rather than 
3) have the debug code still operate on the raw info *anyway*, or by some other approach be able to show the types

I'm pretty sure that we can detect and unwrap Tuple[] stuff just the same as the stuff matching it at runtime does.

**Comment by jjn1056** [comment]
ah ha, if I return message via my gmail it thinks I'm perl-catalyst-sync but via yahoo email I'm jnap.  I guess that's how I setup the sync account.  something to remember.

**Comment by shadowcat-mst** [comment]
Right, but even if we had two way sync working, you'd *still* have to pull the PR down and do it all via command line, because the 'accept PR' button doesn't rebase. If 'accept PR' ever works properly, I'll be happy to revisit two way sync.

**Comment by jjn1056** [comment]
Right PR only works (for a given value of work) if your team drinks all the github koolaide.  I only create a PR because I find it creates a public space for discussion that is open, visible.  Catalyst will always require pulling the branch down and rebasing.

**Comment by jjn1056** [comment]
I'll update a branch based on your proposed 1/2/3 which seems sensible but this time I'm going to add docs to 'list_extra_info' that described these and what they do so that its locked down and we can't screw it up again down the line.

Type::Tiny gives us more than enough meta info to figure out the Tuple and that is the only one I support anyway.  I should add docs to the constraint stuff that specifies that you must use Type::Tiny if you want advanced types since its the only one with a useable interface.

**Comment by shadowcat-mst** [comment]
Even if everybody is using github, you still can't always use the 'accept PR' button. Note that git-hub lets you check out the PR as if it was a branch, so you're still using the PR system.

You just can't accept it, because, even if to begin with everything is sane, if you accept two PRs in a row even if they're unrelated the second one is guaranteed to create a spurious merge commit. The koolade doesn't actually help here, even if everybody uses solely github - it just doesn't work right.

I do, however, very much value the PR as a space for discussion prior to the rebase+merge, and if github ever provides a button for rebase+merge I'd consider bidirectional sync more useful, and in the meantime I'm really glad I can use the */pr/* refs to get a PR down even if the branch was in another repo.

**Comment by shadowcat-mst** [comment]
And, yeah, let's document it and what it's there for, and screw anything except Type::Tiny for this. I can write up a small example somewhere of making an MX::Types library into a Type::Library or something to help people get converted.

**Comment by shadowcat-mst** [comment]
Oh, also - if I seem to've missed something it's because github spams the living bejeesus out of me, so I don't see a lot of the emails. As a backup, I suggest @-ing vanstyn and getting him to kick me up the arse on IRC.

File: lib/Catalyst/DispatchType/Chained.pm [hunk 1]
```diff
@@ -150,13 +162,14 @@ sub list {
             push(@rows, [ '', $name ]);
         }
 
+        my $endpoint_arg_info;
         if($endpoint->has_args_constraints) {
           my $tc = join ',', @{$endpoint->args_constraints};
-          $endpoint .= " ($tc)";
+          $endpoint_arg_info .= " ($tc)";
```
Comment (jjn1056) [inline lib/Catalyst/DispatchType/Chained.pm]: This bit here really clobbered $endpoint, which was totally trouble...

**Comment by jjn1056** [comment]
@shadowcat-mst  worth another review at this point I think.  The parameterized constraints don't seem so bad to me as they stand:

 /chain_base/*/{Tuple[Int,Int,Int]}/*

I'm not against it this way, I did intend the purpose of supported parameterized constraints to be to set the stage for someone to hack in better dependency injection around args, so no reason to not let that stand out in the debug view.  Not sure if this is getting the right level of separation between view and compare logic.  Thoughts?

Added { } around the constraint name, but maybe simple parenthesis is more what Catalyst does in other places.  I went {} since I see that a lot in URI template inspired code, which I adopted for Catalyst::ControllerRole::At.  I don't care that much one way or another, if people prefer something other than {} I'm cool with it.

**Comment by jjn1056** [comment]
@shadowcat-mst (or @vanstyn ) FWIW I finished testing the above on all my personal and work apps and it seems to be fine, even stuff I have that use freaky bits like Catalyst::ActionSignatures and Catalyst::ControllerRole::At ( I stopped testing CatalystX::Declare stuff years ago, I don't think anyone sane used that in prod...)

**Comment by jjn1056** [comment]
@shadowcat-mst @vanstyn  ping :)

**Comment by jjn1056** [comment]
@shadowcat-mst ( @vanstyn ) hey guys, let me know what you think here or if you prefer I can write a distropref to fix this and just go on my own.  This issue makes my logs noisy and I'd like to clear it up.  I don't plan to attend YAPC this year so if you were waiting to talk face to face its going to be a long wait :)

**Comment by shadowcat-mst** [comment]
Looks fine to me - merge and ship at will

**Comment by shadowcat-mst** [comment]
(note that github notifications tend to arrive 50-100 at once for me, no matter how many things I unsubscribe from, so if you want me to see stuff you need to poke me, or ask vanstyn to since he's volunteered to help mitigate my fail here ;)

**Comment by jjn1056** [comment]
I'll try to do it today then, my wife is going in for surgery this week and I'll going too start just working on that and cut out the $work and side coding.  Since you might be paying attention for a second @shadowcat-mst someone I work with said you mentioned Template::Pure and I'd really like to make that not suck as much as it currently does (I mostly like the interface but the source code is sad).  Feel free to aim discourse at me otherwise you might end up seeing a Catalyst::View::Template::Pure that could have been really great but wasn't...

**Comment by jjn1056** [comment]
@shadowcat-mst  yes, if you don't mind dealing with GH issues, over at https://github.com/jjn1056/Template-Pure/issues I already wrote out some stuff, mostly brain dumps that came up while I was actually using this thing on a project.    The main thing is that I feel Zoom as an engine for this would ultimately be better since Zoom has the need streaming and buffer flushing features.  DOM::Tiny has to hold it all in memory and flush all in one go...   But my thinking here was more an extended code meditation to answer the question "I keep saying I want this, but do I really?"  and "If I had this what would I realize I need instead?', hence the warning in the module docs.

FWIW there's a semi working Catalyst View as part of a card game I'm writing for someone's startup, that is ugly but also possibly fun: https://github.com/jjn1056/RangersCCG/blob/master/lib/Catalyst/View/Pure.pm

and example view (with inherited base view) at https://github.com/jjn1056/RangersCCG/blob/master/lib/RangersCCG/Web/View/DisplayStatus.pm

to give you an idea of what this makes me think of.
